### ISSUE TITLE
Performance problems with dense macro models

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#54]
**Current Status:** Unsolved
**Version discovered in:** Python 76, not sure if it goes back yet
**Brief statement of problem:** CV macro models are **Very** slow when run with h20_standard73 atomic data
**How to reproduce:** CV macro model, h20_standard73 data
**Discussion:**

Runs are taking an awfully long time when trying to run CV macro atom models. 

Here is a comparison between the profiler for a QSO run with h20_standard73 atomic data, and a CV run with the same atomic data

Figure 1, CV run:
![screen shot 2013-10-03 at 13 54 34](https://f.cloud.github.com/assets/3007249/1261562/05c5f63e-2c30-11e3-9a30-be17889eec1c.png)

QSO run:

![screen shot 2013-10-03 at 14 31 14](https://f.cloud.github.com/assets/3007249/1261578/717e46a6-2c30-11e3-9c62-6899109dd7b3.png)

In these runs, hydrogen is the only thing being treated as a macro atom. Notice, that the CV run is spending 80% of its time in the matom() routine, i.e. going through macro atom excitations and transitions. 

The reason this happens is because even though the wind is mostly ionsied, there are still densities of 1.0e9 or so in neutral hydrogen. In the QSO case (which we know goes pretty wrong, so take with a pinch of salt) we spend no time at all in the matom routine, which might be wrong, but might be because there is less neutral h in these cases (~1.0e3 , for example).

The bottom line seems to be that in cases where the wind is dense and still has some neutral hydrogen then we seem to get performance problems due to packets having very high line optical depths and so spending a lot of time exciting and de exciting macro atoms.

**Comment by jhmatthews** [comment]
I never got round to coding something up for this because I always used to have plenty of cores in hand. That is no longer the case and performance of the code is slowing my progress, so I've done something about this problem while I'm waiting for some CV runs to finish (not yet fully tested or merged). 

The motivation is that in dense regions we tend to approach LTE in higher levels of macro-atoms and end up with very slow calculations where we constantly make collisional jumps between high up levels- sort of like simulating detailed balance, which is v v slow. The solution is that one should create a superlevel above some threshold and each time you jump into a level inside the superlevel then you choose to jump down from a different one according to LTE populations (divided by statistical weight).

Before, I was struggling to work out how to choose the superlevel threshold, and it seemed hard, but I'd overlooked that you can actually continue to calculate level populations properly in all levels even though you adopt this approximation, and the only difference will be in the distribution of packet paths. This is one of the weird quirks of the macro-atom scheme, where the jumps between levels in the macro-atom only contribute indirectly to the level population calculation (I think it's why the radiation field isn't actually that sensitive to what you assume for level pops, see Lucy). So, I can simply go down through every level each time and work out it's departure coefficient, and if it is within a factor, F, of 1 then flag it as in the superlevel.

I've coded an initial version of this up and it looks to cause a factor ~2 speed up. This could be more like no speed up in less dense models and bigger in dense models. I need to test that it is working properly, but it looks reasonable.

You can choose whether to turn on superlevels in a macro-atom in the atomic data. There's a question about how low you should be allowed to go (i've set a lower floor of level 10), and how to set F. But so far it's looking quite good.

I think the way I'm choosing the level to jump down is right...i.e. I choose my new level j according to a probability P:

```
P = (1/N) n_j* / g_j 
```

where n_j\* is the LTE level population and (1/N) is some normalisation.

**Comment by kslong** [comment]
In a dense CV model with dense atmosphere, the upper levels can be in LTE, and in this case the collisional rate is the highest rate and the packet gets stuck in these upper levels.  James has discussed solutions with Stuart, but these will take some time to implement.  This is currently just a performance issue but it does completely dominate the run time.

**Comment by jhmatthews** [comment]
This issue was discussed in Belfast and Stuart described a solution and a general enhancement of macro-atom method.

This involves replacing the process of jumping through the macro-atom machinery with a method in which a matrix of jumping probabilities is inverted in order to obtain the de-activation state vector, which is mathematically equivalent to a matrix operating n times on an original vector state (the activation state).

I will post references and documents here when they are available. This would be a major improvement of the macro-atom mode in dense regions where jumps are frequent, but would involve a fair amount of careful book-keeping, storing large matrices in memory and doing inversions of large matrices with GSL routines.

**Comment by jhmatthews** [comment]
I'm closing this issue, since there is now a new issue opened specifically to deal with the accelerated matrix solution. This closed issue remains as a document of some of the issues. See #423.
